### PR TITLE
[12.0][MIG] - Migration sale product set 

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

**Review by gva-acsone — APPROVED** [review]
Functional test 👍

File: sale_product_set/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    'category': 'Sale',
+    'license': 'AGPL-3',
+    'author': 'Anybox, Odoo Community Association (OCA)',
+    'version': '12.0.1.0.0',
+    'website': 'https://github.com/OCA/sale-workflow',
+    'summary': "Sale product set",
+    'depends': [
+        'sale',
+    ],
+    'data': [
```
Comment (nikul-serpentcs) [inline sale_product_set/__manifest__.py:15]: Sequence as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

**Comment by CasVissers-360ERP** [comment]
image
I think we should add the "Add Set" option here. Makes a lot more sense.

**Comment by pedrobaeza** [comment]
Good idea!

File: sale_product_set/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2015 Anybox
```
Comment (rven) [inline sale_product_set/__manifest__.py:1]:
```suggestion
```

File: sale_product_set/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright 2018 Camptocamp, ACSONE SA/NV
```
Comment (rven) [inline sale_product_set/__manifest__.py:2]:
```suggestion
```

File: sale_product_set/models/product_set.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2015 Anybox S.A.S
```
Comment (rven) [inline sale_product_set/models/product_set.py:1]:
```suggestion
```

File: sale_product_set/models/product_set.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# Copyright 2016-2018 Camptocamp SA
```
Comment (rven) [inline sale_product_set/models/product_set.py:2]:
```suggestion
```

File: sale_product_set/models/product_set.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import api, fields, models
+
+
+class ProductSet(models.Model):
+    _name = 'product.set'
+    _description = 'Product set'
+
+    name = fields.Char(string='Name', help='Product set name', required=True)
```
Comment (rven) [inline sale_product_set/models/product_set.py]:
```suggestion
    name = fields.Char(help='Product set name', required=True)
```

File: sale_product_set/models/product_set_line.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2015 Anybox S.A.S
```
Comment (rven) [inline sale_product_set/models/product_set_line.py:1]:
```suggestion
```

File: sale_product_set/models/product_set_line.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+# Copyright 2016-2018 Camptocamp SA
```
Comment (rven) [inline sale_product_set/models/product_set_line.py:2]:
```suggestion
```

File: sale_product_set/models/product_set_line.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    _order = 'sequence'
+
+    product_id = fields.Many2one(
+        comodel_name='product.product',
+        domain=[('sale_ok', '=', True)],
+        string='Product',
+        required=True
+    )
+    quantity = fields.Float(
+        string='Quantity',
```
Comment (rven) [inline sale_product_set/models/product_set_line.py:21]:
```suggestion
```

File: sale_product_set/models/product_set_line.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        required=True,
+        default=1
+    )
+    product_set_id = fields.Many2one(
+        'product.set',
+        string='Set',
+        ondelete='cascade',
+    )
+    sequence = fields.Integer(
+        string='Sequence',
```
Comment (rven) [inline sale_product_set/models/product_set_line.py:32]:
```suggestion
```

File: sale_product_set/tests/test_product_set.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2015 Anybox S.A.S
```
Comment (rven) [inline sale_product_set/tests/test_product_set.py:1]:
```suggestion
```

File: sale_product_set/tests/test_product_set.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+# Copyright 2016-2018 Camptocamp SA
```
Comment (rven) [inline sale_product_set/tests/test_product_set.py:2]:
```suggestion
```

File: sale_product_set/wizard/product_set_add.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2015 Anybox S.A.S
```
Comment (rven) [inline sale_product_set/wizard/product_set_add.py:1]:
```suggestion
```

File: sale_product_set/wizard/product_set_add.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+# Copyright 2016-2018 Camptocamp SA
```
Comment (rven) [inline sale_product_set/wizard/product_set_add.py:2]:
```suggestion
```

File: sale_product_set/wizard/product_set_add.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+class ProductSetAdd(models.TransientModel):
+    _name = 'product.set.add'
+    _rec_name = 'product_set_id'
+    _description = "Wizard model to add product set into a quotation"
+
+    product_set_id = fields.Many2one(
+        'product.set', 'Product set', required=True)
+    quantity = fields.Float(
+        string='Quantity',
```
Comment (rven) [inline sale_product_set/wizard/product_set_add.py]:
```suggestion
```

**Review by rven — CHANGES_REQUESTED** [review]
Hi @sbejaoui,

Can you have a look at my code improvements?

File: sale_product_set/__manifest__.py (hunk 1 above)
Comment (sbejaoui) [reply sale_product_set/__manifest__.py:15]: done

**Comment by sbejaoui** [comment]
Hi @rven ,
Why would you remove the copyrights from .py source file?

**Comment by sbejaoui** [comment]
> image
> I think we should add the "Add Set" option here. Makes a lot more sense.

@CasVissers ,@pedrobaeza ,

I gave it a try, but the way this controls are developed don't help:

- The widget is defined [here](https://github.com/odoo/odoo/blob/12.0/addons/account/static/src/js/section_and_note_fields_backend.js), in the account module (I don't get why). That's make unnecessary dependency. IMO, this widget should be in the web module.
- I don't see a clean way to inherit the [`SectionAndNoteListRenderer`](https://github.com/odoo/odoo/blob/12.0/addons/account/static/src/js/section_and_note_fields_backend.js#L18) as the widget does not return it.

**Review by sbidoul — APPROVED** [review]

**Comment by rven** [comment]
@sbejaoui I had to do it also on another PR, but it is only needed for __init__.py files.
So you can ignore these comments of me.

**Comment by sbidoul** [comment]
@rven can you update your review so we can merge this?

**Review by rven — APPROVED** [review]

**Review by CasVissers-360ERP — APPROVED** [review]

**Comment by nikul-serpentcs** [comment]
@sbejaoui Use new README by fragment system, copying https://github.com/OCA/maintainer-tools/tree/master/template/module/readme and editing or removing the needed sections. You can then run locally oca-gen-addon-readme if you have OCA/maintainer-tools installed and you want to pre-generate the README.rst file (preferred), or keep the previous README.rst file.

**Comment by sbejaoui** [comment]
@nikul-serpentcs , it's already done

**Comment by nikul-serpentcs** [comment]
> @nikul-serpentcs , it's already done

@sbejaoui But it does not look like this use ` oca-gen-addon-readme`

**Comment by sbidoul** [comment]
@nikul-serpentcs the readme will be generated automatically after merging.

**Comment by pedrobaeza** [comment]
@sbidoul please next time fix commit history before merging. And remember to check the mark in the migration issue.

**Pull request merged**